### PR TITLE
feature: Add the serilog target

### PR DESCRIPTION
**Pull request by glennawatson**
The credit for this goes to @joelweiss 

Adding Serilog support to Splat directly as discussed in https://github.com/joelweiss/Splat.Serilog/issues/3

This follows the model of the other couple loggers we have.

### CODE REVIEW

File: src/Splat.Serilog/MutableDependencyResolverExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) 2019 .NET Foundation and Contributors. All rights reserved.
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for full license information.
+
+namespace Splat.Serilog
+{
+    /// <summary>
+    /// NLog specific extensions for the Mutable Dependency Resolver.
```
Comment (joelweiss) [inline src/Splat.Serilog/MutableDependencyResolverExtensions.cs]:
```suggestion
    /// Serilog specific extensions for the Mutable Dependency Resolver.
```

File: src/Splat.Serilog/MutableDependencyResolverExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /// </summary>
+    public static class MutableDependencyResolverExtensions
+    {
+        /// <summary>
+        /// Simple helper to initialize NLog within Splat with the Wrapping Full Logger.
```
Comment (joelweiss) [inline src/Splat.Serilog/MutableDependencyResolverExtensions.cs]:
```suggestion
        /// Simple helper to initialize Serilog within Splat with the Wrapping Full Logger.
```

File: src/Splat.Serilog/MutableDependencyResolverExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        /// <remarks>
+        /// You should configure NLog prior to calling this method.
```
Comment (joelweiss) [inline src/Splat.Serilog/MutableDependencyResolverExtensions.cs]:
```suggestion
        /// You should configure Serilog prior to calling this method.
```

File: src/Splat.Serilog/MutableDependencyResolverExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        /// </remarks>
+        /// <param name="instance">
+        /// An instance of Mutable Dependency Resolver.
+        /// </param>
+        /// <example>
+        /// <code>
+        /// Locator.CurrentMutable.UseNLogWithWrappingFullLogger();
```
Comment (joelweiss) [inline src/Splat.Serilog/MutableDependencyResolverExtensions.cs]:
```suggestion
        /// Locator.CurrentMutable.UseSerilogWithWrappingFullLogger();
```

File: src/Splat.Serilog/SerilogLogger.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            new KeyValuePair<LogLevel, LogEventLevel>(LogLevel.Fatal, LogEventLevel.Fatal)
+        };
+
+        private readonly global::Serilog.ILogger _inner;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SerilogLogger"/> class.
+        /// </summary>
+        /// <param name="inner">The actual serilog logger.</param>
+        /// <exception cref="ArgumentNullException">NLog logger not passed.</exception>
```
Comment (joelweiss) [inline src/Splat.Serilog/SerilogLogger.cs]:
```suggestion
        /// <exception cref="ArgumentNullException">Serilog logger not passed.</exception>
```

**Comment by joelweiss** [comment]
We should also have a way to configure with a logger instance

```csharp
public static void UseSerilogWithWrappingFullLogger(this IMutableDependencyResolver instance, global::Serilog.ILogger logger)
```

**Comment by glennawatson** [comment]
@joelweiss With deprecated API our usual project policy on RxUI based projects is to keep them around for one major revision. Eg we'll keep the Registrations around until v8 which is usually around 1-6 timeline. Depends when we do a breaking API change since we follow semver.org versioning.

**Comment by joelweiss** [comment]
After taking a closer look I still have a few notes, let me know if you want me to raise it in a separate issue/pr

* calling `Locator.CurrentMutable.UseSerilogWithWrappingFullLogger(logger)` will throw an `InvalidCastException`.
<details>
 <summary>Tests</summary>

```csharp
/// <summary>
/// Test to make sure the calling `UseSerilogWithWrappingFullLogger` logs.
/// </summary>
[Fact]
public void Configuring_With_Static_Log_Should_Write_Message()
{
	var serilogLoggerAndTarget = GetActualSerilogLoggerAndTarget();
	Log.Logger = serilogLoggerAndTarget.Logger;
	var target = serilogLoggerAndTarget.Target;
	Locator.CurrentMutable.UseSerilogWithWrappingFullLogger();

	Assert.Equal(0, target.Logs.Count);

	IEnableLogger logger = null;
	logger.Log().Debug<DummyObjectClass2>("This is a test.");

	Assert.Equal(1, target.Logs.Count);
	Assert.Equal($"{nameof(DummyObjectClass2)}: This is a test.", target.Logs.First());
}

/// <summary>
/// Test to make calling `UseSerilogWithWrappingFullLogger(Serilog.ILogger)` logs.
/// </summary>
[Fact]
public void Configuring_With_PreConfigured_Log_Should_Write_Message()
{
	var serilogLoggerAndTarget = GetActualSerilogLoggerAndTarget();
	var target = serilogLoggerAndTarget.Target;
	Locator.CurrentMutable.UseSerilogWithWrappingFullLogger(serilogLoggerAndTarget.Logger);

	Assert.Equal(0, target.Logs.Count);

	IEnableLogger logger = null;
        //Will Fail
	logger.Log().Debug<DummyObjectClass2>("This is a test.");

	Assert.Equal(1, target.Logs.Count);
	Assert.Equal($"{nameof(DummyObjectClass2)}: This is a test.", target.Logs.First());
}
```
</details>
<details>
 <summary>Suggested Fix:</summary>

```csharp
public static void UseSerilogWithWrappingFullLogger(this IMutableDependencyResolver instance)
{
	instance.UseSerilogWithWrappingFullLogger(global::Serilog.Log.Logger);
}

public static void UseSerilogWithWrappingFullLogger(this IMutableDependencyResolver instance, global::Serilog.ILogger actualLogger)
{
	var funcLogManager = new FuncLogManager(type =>
	{
		var miniLoggingWrapper = new SerilogLogger(actualLogger.ForContext(type));
		return new WrappingFullLogger(miniLoggingWrapper);
	});

	instance.RegisterConstant(funcLogManager, typeof(ILogManager));
}
```
</details>
</br>

* The whole idea of serilog is to use structured logging, and for that to work we have to pass in the format string ("Event {Foo} has happened") and the arguments to the logger, and by using [`WrappingFullLogger`](https://github.com/reactiveui/splat/blob/d4881ddfd69dd3f9c798e71fa10c6cb6e3b241aa/src/Splat.Serilog/MutableDependencyResolverExtensions.cs#L31) which uses `string.Format()` on all log events we lose that.


* instead of 
```chsarp
_inner.Write(SplatLogLevelToSerilogLevel(logLevel), exception, $"{type.Name}: {message}");
```
it should be
```chsarp
_inner.ForContext(type).Write(SplatLogLevelToSerilogLevel(logLevel), message);
```

**Comment by glennawatson** [comment]
Raise in a separate issue anyway. Closed PR have a habit of getting lost

**Pull request merged**